src/conn.ts

```diff
@@ -23,106 +28,364 @@ export class ConnOptions {
-              this.bot.recipes = [...this.bot.recipes, ...data.recipes1];
-              break;
-            case 2: //* remove
-              this.bot.recipes = Array.from(
-                (data.recipes1 as number[]).reduce((recipes, recipe) => {
-                  recipes.delete(recipe);
-                  return recipes;
-                }, new Set(this.bot.recipes))
-              );
-              break;
+    if (options?.toClientMiddleware) this.toClientDefaultMiddleware = options.toClientMiddleware;
+    if (options?.toServerMiddleware) this.toServerDefaultMiddleware = options.toServerMiddleware;
+
+    this.bot._client.on('raw', this.onServerRaw.bind(this))
+  }
+
+  /**
+   * Called when the proxy bot receives a packet from the server. Forwards the packet to all attached and receiving clients taking
+   * attached middleware's into account.
+   * @param buffer Buffer
+   * @param meta 
+   * @returns 
+   */
+  async onServerRaw(buffer: Buffer, meta: PacketMeta) {
+    // @ts-ignore-error
+    const packetData = this.bot._client.deserializer.parsePacketBuffer(buffer).data.params
+    for (const pclient of this.receivingPclients) {
+      if (pclient.state !== states.PLAY || meta.state !== states.PLAY) { return }
+      // Build packet canceler function used by middleware
+      const cancel: packetCanceler = Object.assign((unCancel: boolean = false) => {
+        if (unCancel === true) {
+          cancel.isCanceled = false
+        } else {
+          cancel.isCanceled = true
+        }
+        update.isUpdated = true
+      }, { isCanceled: false })
+      const update: packetUpdater = Object.assign((unUpdate: boolean = false) => {
+        if (unUpdate === false) {
+          update.isUpdated = false
+        }
+        update.isUpdated = true
+      }, { isUpdated: false })
+
+      for (const middleware of pclient.toClientMiddlewares) {
+        const funcReturn = middleware({ bound: 'client', meta, writeType: 'packet' }, pclient, packetData, cancel, update)
+        if (funcReturn instanceof Promise) {
+          await funcReturn
+        }
+      }
+      // TODO: figure out what packet is breaking crafting on 2b2t
+      // Hint: It is the recipes unlock packet that is send when crafting an item. 
+      // Probably some bad unlocked recipes packet reconstruction on login that is causing packets send after to crash the client.
+      
+      if (cancel.isCanceled === false) {
+        if (!update.isUpdated && !writeBuffer) {
+          pclient.writeRaw(buffer)
+          return
+        }
+        // pclient.write(meta.name, data);
+        if (writeBuffer) {
+          // @ts-ignore-error
+          const packetBuff = pclient.serializer.createPacketBuffer({ name: meta.name, params: packetData })
+          if (!bufferEqual(buffer, packetBuff)) {
+            console.log('client<-server: Error in packet ' + meta.state + '.' + meta.name + ' This packet is breaking the proxy. Ping @Ic3Tank on NextGen discord')
```

Comment:
```suggestion
            console.error(`client<-server: Error in packet ${meta.state}.${meta.name} Issue Reserializing, please open an issue or contact the maintainer(s) of mcproxy directly.`)
```
maybe a more generalized error message which is actually printed to stderr,
though rather just not have it at all